File: src/main/scala/g0001_0100/s0024_swap_nodes_in_pairs/Solution.scala
package g0001_0100.s0024_swap_nodes_in_pairs

// #Medium #Top_100_Liked_Questions #Linked_List #Recursion #Data_Structure_II_Day_12_Linked_List
// #Udemy_Linked_List #Big_O_Time_O(n)_Space_O(1)
// #2023_10_30_Time_441_ms_(95.65%)_Space_56.9_MB_(52.17%)

import com_github_leetcode.ListNode

/*
 * Definition for singly-linked list.
 * class ListNode(_x: Int = 0, _next: ListNode = null) {
 *   var next: ListNode = _next
 *   var x: Int = _x
 * }
 */
object Solution {
    def swapPairs(head: ListNode): ListNode = {
        if (head == null) {
            return null
        }
        val len = getLength(head)
        reverse(head, len)
    }

    private def getLength(curr: ListNode): Int = {
        var cnt = 0
        var current = curr
        while (current != null) {
            cnt += 1
            current = current.next
        }
        cnt
    }

    private def reverse(head: ListNode, len: Int): ListNode = {
        if (len < 2) {
            return head
        }
        var curr = head
        var prev: ListNode = null
        var next: ListNode = null
        for (i <- 0 until 2) {
            next = curr.next
            curr.next = prev
            prev = curr
            curr = next
        }
        head.next = reverse(curr, len - 2)
        prev
    }
}

File: src/main/js/g0001_0100/s0024_swap_nodes_in_pairs/solution.js
// #Medium #Top_100_Liked_Questions #Linked_List #Recursion #Data_Structure_II_Day_12_Linked_List
// #Udemy_Linked_List #Big_O_Time_O(n)_Space_O(1)
// #2024_12_04_Time_1_ms_(100.00%)_Space_49.6_MB_(8.69%)

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
const swapPairs = function (head) {
    if (head === null) {
        return null
    }

    const getLength = (node) => {
        let count = 0
        while (node !== null) {
            count++
            node = node.next
        }
        return count
    };

    const reverse = (node, length) => {
        if (length < 2) {
            return node
        }

        let current = node
        let prev = null
        let next

        for (let i = 0; i < 2; i++) {
            next = current.next
            current.next = prev
            prev = current
            current = next
        }

        node.next = reverse(current, length - 2)
        return prev
    };

    const len = getLength(head)
    return reverse(head, len)
};

export { swapPairs }
